fw/http_parser.c

```diff
@@ -3558,6 +3584,449 @@ __req_parse_x_forwarded_for(TfwHttpMsg *hm, unsigned char *data, size_t len)
+} while (0)
+
+/* 
+ * Tries to find parameter in header.
+ * Parsing fails if parameter not a unique in current header.
+ *
+ * RFC 7239 section 4: 
+ * Each parameter MUST NOT occur more than once per field-value.
+ */
+#define FWD_TRY_STR_NAME(name, curr_st, next_st, fwd_flag)		\
+	TRY_STR_LAMBDA_fixup_flag(&TFW_STR_STRING(name),		\
+				  &parser->hdr, { 			\
+				  if (parser->flags & fwd_flag)	\
+					return CSTR_NEQ;		\
+				  parser->flags |= fwd_flag;	\
+				  }, curr_st, next_st,			\
+				  TFW_STR_NAME)
+
+	__FSM_START(parser->_i_st);
+
+	__FSM_STATE(Req_I_Fwd) {
+		FWD_TRY_STR_NAME("for=", Req_I_Fwd, Req_I_Fwd_For_Start,
+				 FWD_SET_FOR);
+		FWD_TRY_STR_NAME("host=", Req_I_Fwd, Req_I_Fwd_Host_Start,
+				 FWD_SET_HOST);
+		FWD_TRY_STR_NAME("proto=", Req_I_Fwd, Req_I_Fwd_Proto_Start,
+				 FWD_SET_PROTO);
+		FWD_TRY_STR_NAME("by=", Req_I_Fwd, Req_I_Fwd_By_Start,
+				 FWD_SET_BY);
+		return CSTR_NEQ;
+	}
+
+	__FSM_STATE(Req_I_Fwd_For_List) {
+		/* Eat OWS before parameter. */
+		if (unlikely(IS_WS(c)))
+			__FSM_I_MOVE_fixup(Req_I_Fwd_For_List, 1, 0);
+		/* Find next "for=" in list. */
+		FWD_TRY_STR_NAME("for=",
+				 Req_I_Fwd_For_List,
+				 Req_I_Fwd_For_Start,
+				 0);
+		return CSTR_NEQ;
+	}
+
+	__FSM_STATE(Req_I_Fwd_For_Start) {
+		if (likely(c == '"'))
+			__FSM_I_MOVE_fixup(Req_I_Fwd_For_Quoted, 1, 0);
+		/* Fall through */
+	}
+
+	__FSM_STATE(Req_I_Fwd_For_Unquoted) {
+		/*
+		 * Eat IP address or host name.
+		 *
+		 * TODO: parse/validate IP addresses and textual IDs.
+		 * Currently we just validate separate characters, but the
+		 * whole value may be invalid (e.g. "---[_..[[").
+		 */
+		__FSM_I_MATCH_MOVE_fixup(xff, Req_I_Fwd_For_Node_Id_Unquoted,
+					 TFW_STR_VALUE);
+		if (unlikely(!__fsm_sz))
+			return CSTR_NEQ;
+		__FSM_I_MOVE_fixup(Req_I_Fwd_For_Sep, __fsm_sz, TFW_STR_VALUE);
+	}
+
+	/*
+	 * At this state we know that we saw at least one character as
+	 * a host address and now we can pass zero length token.
+	 */
+	__FSM_STATE(Req_I_Fwd_For_Node_Id_Unquoted) {
+		__FSM_I_MATCH_MOVE_fixup(xff,
+					 Req_I_Fwd_For_Node_Id_Unquoted,
+					 TFW_STR_VALUE);
+		__FSM_I_MOVE_fixup(Req_I_Fwd_For_Sep, __fsm_sz, TFW_STR_VALUE);
+	}
+
+	__FSM_STATE(Req_I_Fwd_For_Quoted) {
+		__FSM_I_MATCH_MOVE_fixup(xff,
+					 Req_I_Fwd_For_Node_Id_Quoted,
+					 TFW_STR_VALUE);
+		if (unlikely(!__fsm_sz))
+			return CSTR_NEQ;
+		__FSM_I_MOVE_fixup(Req_I_Fwd_For_Sep_Quoted, __fsm_sz,
+				   TFW_STR_VALUE);
+	}
+
+	__FSM_STATE(Req_I_Fwd_For_Node_Id_Quoted) {
+		__FSM_I_MATCH_MOVE_fixup(xff,
+					 Req_I_Fwd_For_Node_Id_Quoted,
+					 TFW_STR_VALUE);
+		__FSM_I_MOVE_fixup(Req_I_Fwd_For_Sep_Quoted, __fsm_sz,
+				   TFW_STR_VALUE);
+	}
+
+	__FSM_STATE(Req_I_Fwd_For_Sep_Quoted) {
+		/* At this point we tries fixup '"' with near symbol */
+
+		if (unlikely(c != '"'))
+			return CSTR_NEQ;
+		/* End of data on quote */
+		if (unlikely(__data_off(p + 1) >= len))
+			__FSM_I_MOVE_fixup(Req_I_Fwd_For_Sep, 1, 0);
```

Comment:
IIUC you suggest to remove part of `Req_I_Fwd_For_Sep_Quoted` and instead use aggregated `Req_I_Fwd_For_Sep` and `Req_I_Fwd_For_Sep_End`. However, `Req_I_Fwd_For_Sep_Quoted` need to fixup quote in one chunk with another near symbol. E.g  ' "; ' or ' ", '. If I use `__FSM_I_MOVE(Req_I_Fwd_For_Sep)` quote which now at @p, will be skipped.

Reply:
So we need to fixup `"` with the next chracter, right? Can we just use explicit fixup, like with `__msg_field_open()`?

Reply:
We have few explicit fixup functions. Some of them:
`__msg_field_open()` Initialize TfwStr for subsequent using. 
`__msg_field_fixup()` Just regular field fixup.
`__msg_field_finish()` Fixup field and mark `TfwStr` as completed.

`__msg_field_open()` useful when we have empty `TfwStr` and we need to initialize it for using with fixup functions.

`__msg_field_open()` in fact is:
```
	str->data = data;
	str->skb = skb;
```
it's just preparation of `TfwStr`(actually just a chunk) for later using in `__tfw_http_msg_add_str_data()`.

`__tfw_http_msg_add_str_data()` - Main fixup logic. If passed empty `TfwStr`, we just update length of current string, because it is not compound. Otherwise, allocate chunk, then call `__tfw_http_msg_set_str_data()` which essentially `__msg_field_open()`  and calculate new length of `TfwStr`.

As we can see, `__msg_field_open()` doesn't allocate chunks, therefore we must do it manually.
One case come to mind: Use simple  `__FSM_I_MOVE_fixup`, then just update length of new chunk. But it's not correct approach. If we will use `tfw_str_updlen`to update length of chunk, we got assertion. `BUG_ON(chunk->len);`

But there is another approach that should work(tests passed), but it's overcomplicated, although have one benefit, even when request postponed next symbol after quote fixuped:
```
	__FSM_STATE(Req_I_Fwd_For_Sep_Quoted) {
		TfwStr *sn = NULL;

		if (unlikely(c != '"'))
			return CSTR_NEQ;

		sn = tfw_str_add_compound(hm->pool, &parser->hdr);

		if (!sn) {
			T_WARN("Cannot grow HTTP data string\n");
			return -ENOMEM;
		}
		__msg_field_open(sn, p);

		/* Why we just don't use __FSM_I_MOVE? Because
		 * at the end of fixup, length will be calcculated
		 * and then we'll have assertion in tfw_str_updlen.
		 */
		if (__data_off(p + 1) >= len) {
			parser->_i_st = &&Req_I_Fwd_For_Sep_Quoted_n;
			return TFW_POSTPONE;
		}
		p += 1;
	}

	__FSM_STATE(Req_I_Fwd_For_Sep_Quoted_n) {
		/* At this point we try to fixup '"' with near symbol. */

		/* EOL after quote */
		if (likely(IS_CRLF(c))) {
			tfw_str_updlen(&parser->hdr, p);
			return __data_off(p);
		}
		/* ';' after quote */
		if (likely(c == ';')) {
			tfw_str_updlen(&parser->hdr, p);
			__FSM_I_MOVE(Req_I_Fwd);
		}
		/* ',' after quote */
		if (unlikely(c == ',')) {
			tfw_str_updlen(&parser->hdr, p);
			__FSM_I_MOVE(Req_I_Fwd_For_List);
		}
		/* WS after quote */
		if (unlikely(IS_WS(c))) {
			tfw_str_updlen(&parser->hdr, p);
			__FSM_I_MOVE(Req_I_Fwd_For_Sep_End);
		}

		return CSTR_NEQ;
	}
```
Few words about `__msg_field_open()` comments.
```
 * 2. call __msg_field_open() to initialize a new _empty_ TfwStr chunk (see
 *    __tfw_http_msg_add_str_data())
 ```
Looks like typo.  Is `tfw_http_msg_set_str_data` implied?

Reply:
I still don't understand the problem and probably we need one more discussion on the code...

Yes, the comment should be fixed with `tfw_http_msg_set_str_data` instead of `__tfw_http_msg_add_str_data`.

I didn't get why (pseudo-)code like
```C
__FSM_STATE(Req_I_Fwd_For_Sep_Quoted) {
    __msg_field_open(sn, p);
    __FSM_I_MOVE(Req_I_Fwd_For_Sep);
    ...
}
```
fails? `__msg_field_open()` initializes the `TfwStr` and `__tfw_http_msg_add_str_data()` from `__FSM_I_MOVE()` goes to the `TFW_STR_EMPTY(str)` branch in `__tfw_http_msg_add_str_data()` and there is no `tfw_str_updlen()` call. 

The code
```C
if (unlikely(__data_off(p + 1) >= len)) {
    __msg_hdr_chunk_fixup(p, 1);
    parser->_i_st = &&Req_I_Fwd_For_Sep;
    return TFW_POSPONE;
}
```
is essentially `__FSM_I_MOVE(Req_I_Fwd_For_Sep)`.

And yes, the comment `/* End of data on quote */` at least must be extended to describe well why not to use `__FSM_I_MOVE` and why do we have all the `p + 1` logic.

Reply:
Yes, you're right: we use `__msg_field_open()` for a new field/header, not jut a chunk of the header. Please add this to the big comment in line 333.

And yes, we need an ad-hoc macro to open a new chunk of the header, fixup it if not enough data here and move to a new state to avoid the `p + 1` stuff in the `Req_I_Fwd_For_Sep_Quoted`.

Please collect and distill all the discussion into a comment for the macro, probably reference it from the big comment at line 333. Once we have similar complex header, we'll probably need to generalize the macro.